silq/pulses/pulse_sequences.py

```diff
@@ -31,11 +37,626 @@ def __setattr__(self, key, value):
+       pulse is defined in ``ESRPulseSequence.ESR['pulse_delay']``.
+    4. Perform read pulse ``ESRPulseSequence.ESR['read_pulse']``.
+    5. Repeat steps 2 and 3 for each ESR pulse in
+       ``ESRPulseSequence.ESR['ESR_pulses']``, which by default contains single
+       pulse ``ESRPulseSequence.ESR['ESR_pulse']``.
+    6. Perform empty-plunge-read sequence (EPR), but only if
+       ``ESRPulseSequence.EPR['enabled']`` is True.
+       EPR pulses are defined in ``ESRPulseSequence.EPR['pulses']``.
+    7. Perform any post_pulses defined in ``ESRPulseSequence.post_pulses``.
+
+    Parameters:
+        ESR (dict): Pulse settings for the ESR part of the pulse sequence.
+            Contains the following items:
+
+            * ``stage_pulse`` (Pulse): Stage pulse in which to perform ESR
+              (e.g. plunge). Default is 'plunge `DCPulse`.
+            * ``ESR_pulse`` (Pulse): Default ESR pulse to use.
+              Default is 'ESR' ``SinePulse``.
+            * ``ESR_pulses`` (List[Union[str, Pulse]]): List of ESR pulses to
+              use. Can be strings, in which case the string should be an item in
+              ``ESR`` whose value is a `Pulse`.
+            * ``pulse_delay`` (float): ESR pulse delay after beginning of stage
+              pulse. Default is 5 ms.
+            * ``read_pulse`` (Pulse): Pulse after stage pulse for readout and
+              initialization of electron. Default is 'read_initialize`
+              `DCPulse`.
+
+        EPR (dict): Pulse settings for the empty-plunge-read (EPR) part of the
+            pulse sequence. This part is optional, and is used for non-ESR
+            contast, and to measure dark counts and hence ESR contrast.
+            Contains the following items:
+
+            * ``enabled`` (bool): Enable EPR sequence.
+            * ``pulses`` (List[Pulse]): List of pulses for EPR sequence.
+              Default is ``empty``, ``plunge``, ``read_long`` `DCPulse`.
+
+        pre_pulses (List[Pulse]): Pulses before main pulse sequence.
+            Empty by default.
+        post_pulses (List[Pulse]): Pulses after main pulse sequence.
+            Empty by default.
+        pulse_settings (dict): Dict containing all pulse settings.
+        **kwargs: Additional kwargs to `PulseSequence`.
+
+    Examples:
+        The following code measures two ESR frequencies and performs an EPR
+        from which the contrast can be determined for each ESR frequency:
+
+        >>> ESR_pulse_sequence = ESRPulseSequence()
+        >>> ESR_pulse_sequence.ESR['pulse_delay'] = 5e-3
+        >>> ESR_pulse_sequence.ESR['stage_pulse'] = DCPulse['plunge']
+        >>> ESR_pulse_sequence.ESR['ESR_pulse'] = FrequencyRampPulse('ESR_adiabatic')
+        >>> ESR_pulse_sequence.ESR_frequencies = [39e9, 39.1e9]
+        >>> ESR_pulse_sequence.EPR['enabled'] = True
+        >>> ESR_pulse_sequence.pulse_sequence.generate()
+
+        The total pulse sequence is plunge-read-plunge-read-empty-plunge-read
```

Comment:
same here

Reply:
will resolve this, we can keep dicussing in comment above